### ISSUE TITLE
Add a checker for expanding list during for iteration

### ISSUE THREAD

**Issue by orSolocate** [PyCQA/pylint#5348]
### Current problem

I would like Pylint to address a classic python bug using `for` loops, when iterating through an expanding list or a dictionary:

For example running:
```py
 item_list = [1, 2, 3]
 for item in item_list:
      item_list.append(1)
     print(item)
```

results in an infinite loop.

For dictionary we have a runtime error raised when this happens:

```py
my_dict = {"1": 1, "2": 2, "3": 3}
i = 1
for item in my_dict:
    my_dict[i] = 1
    i += 1
    print(i)
```

results in: `RuntimeError: dictionary changed size during iteration`

Pylint does not detect both errors.

### Desired solution

I suggest a new checker that detects this structure and suggests using a copy of the list.
in the list example: change `for item in item_list:`  to  `for item in item_list.copy():`
with a message like:
` Expanding list during iteration, consider using <list_name>.copy() instead`
and a name like:
`iteration_expanding_list`

If the community agrees I could implement it :)

### Additional context

See MR #5628

_No response_

**Comment by orSolocate** [comment]
@Pierre-Sassoulas  - is it a Do or a No Do for you guys?

**Comment by Pierre-Sassoulas** [comment]
@DanielNoord added a thumbs up, so I guess he agree with it :) 

I don't remember why I added the discussion label, I agree with the general idea of the checker. Maybe I thought we'd need to discuss the details. What do you think @cdce8p ?

**Comment by cdce8p** [comment]
It looks like a check that could be quite useful! We just need to be careful to not create false positives when iterating over a copy of the list / dict. `.copy()` is the obvious one, but there are others, too. Eg. `copy(d)` or even something like `[k for k in d]`.

The check could also cover `sets` as they also raise a `RuntimeError` in this case.

**Comment by orSolocate** [comment]
@cdce8p  I agree with expanding it to sets too, good idea. 
Do you know what is the take of `astrroid` for this? i.e. for a code like this:
`lst1 = [1]

lst2 = lst1.copy()`

I guess those lists would be considered as different objects. So the algorithm could be simple as this:
```
for any `for` node with a list as it's variable after the `in` word:
raise message if there is an append statement inside the for loop scope regarding this specific variable.
```

**Comment by orSolocate** [comment]
@cdce8p what code produces the error for Sets for you?

running:

`
item_set = {1, 2, 3}
for item in item_set:
    item_set.add(item)
`

in Python 3.7.7 has no runtime error. The runtime error only appears with Dictionaries..

**Comment by orSolocate** [comment]
@cdce8p  How about a shrinking list?
I think it is also an issue to consider, because when you iterate through a list whose items are being deleted the grabbing of the next item is unexpected.
[because the call to `next` is after the item deletion, and if you deleted the current item than you skip the next one(!)]. 
Like in this example:

```
 item_list = [1, 2, 3]
 for item in item_list:
      item_list.remove(item)
      print(item)
```

output:
1
3

explanation: `1` is removed and printed. `next` is called for next loop iteration but since `1` is gone, `next` returns `3`(!), we skipped `2` and it is a difficult bug for the programmer to find without careful debugging.
